### PR TITLE
opt_merge: discard \init of '$' cells with 'Q' port when merging

### PR DESCRIPTION
**Pull request by eddiehung**

### CODE REVIEW

File: passes/opt/opt_merge.cc [hunk 1]
```diff
@@ -222,7 +222,8 @@ struct OptMergeWorker
 			return true;
 		}
 
-		if (cell1->type.begins_with("$") && conn1.count(ID(Q)) != 0) {
+		if (conn1.count(ID(Q)) != 0 && (cell1->type.begins_with("$dff") || cell1->type.begins_with("$dlatch") || cell1->type.in("$adff", "$sr", "$ff") ||
+					cell1->type.begins_with("$_DFF") || cell1->type.begins_with("$_DLATCH"))) {
```
Comment (mwkmwkmwk) [inline passes/opt/opt_merge.cc]: There's also `$_FF_` and `$_SR_*`

File: passes/opt/opt_merge.cc [hunk 2]
```diff
@@ -222,7 +222,8 @@ struct OptMergeWorker
 			return true;
 		}
 
-		if (cell1->type.begins_with("$") && conn1.count(ID(Q)) != 0) {
```
Comment (mwkmwkmwk) [inline passes/opt/opt_merge.cc]: Awww, that was a perfect "featherless biped"-style definition :(

File: passes/opt/opt_merge.cc (hunk 1 above)
Comment (eddiehung) [reply passes/opt/opt_merge.cc]: In a future state element refactor, I propose that we add a `bool cell_ff(IdString type)` function (amongst others) to: https://github.com/YosysHQ/yosys/blob/319cba70d37eafdb8bbd3ddf6a0f9c238d53d0c2/kernel/celltypes.h#L34

Comment (whitequark) [reply passes/opt/opt_merge.cc]: See also: https://github.com/YosysHQ/yosys/pull/1562/files#diff-94b3e6687aa5c05fbb80cfdfba068adcR173-R198

**Comment by nakengelhardt** [comment]
@eddiehung I think there was no problem with this PR but it was forgotten over the holidays. Can you fix the conflict and merge this?

**Comment by eddiehung** [comment]
> @eddiehung I think there was no problem with this PR but it was forgotten over the holidays. Can you fix the conflict and merge this?

Done. Sorry for mixing in some unrelated changes which I've backed out -- they were the cause of the conflict in the first place.

**Pull request merged**